File: frontend/src/types/userLibrary.ts
export interface UserLibrarySavedItem {
  news_id: number;
  marked_at: number;
}

export interface UserLibraryReadItem {
  news_id: number;
  read_at: number;
}

export interface UserLibrarySnapshot {
  saved: UserLibrarySavedItem[];
  read: UserLibraryReadItem[];
}

export interface UserLibrarySavedDelta {
  news_id: number;
  marked_at: number;
  removed: boolean;
}

export interface UserLibraryPutRequest {
  saved: UserLibrarySavedDelta[];
  read: UserLibraryReadItem[];
}

export interface UserLibraryPutResponse {
  saved_count: number;
  read_count: number;
  skipped_news_ids: number[];
}

File: backend/app/schemas/user_library.py
"""Request/response payloads for ``/users/me/library``."""

from __future__ import annotations

from pydantic import BaseModel, Field


class LibrarySavedItem(BaseModel):
    news_id: int = Field(gt=0)
    marked_at: int = Field(gt=0, description="Unix time in milliseconds")


class LibraryReadItem(BaseModel):
    news_id: int = Field(gt=0)
    read_at: int = Field(gt=0, description="Unix time in milliseconds")


class LibrarySavedDelta(BaseModel):
    news_id: int = Field(gt=0)
    marked_at: int = Field(
        gt=0,
        description="Event time in milliseconds: mark time, or unmark time when removed is true",
    )
    removed: bool = False


class LibraryGetResponse(BaseModel):
    saved: list[LibrarySavedItem]
    read: list[LibraryReadItem]


class LibraryPutRequest(BaseModel):
    saved: list[LibrarySavedDelta] = Field(default_factory=list, max_length=2000)
    read: list[LibraryReadItem] = Field(default_factory=list, max_length=2000)

    model_config = {"extra": "forbid"}


class LibraryPutResponse(BaseModel):
    saved_count: int
    read_count: int
    skipped_news_ids: list[int]
